RFC - Type extension and duplication (#428)
blockprotocol/blockprotocol: rfcs/text/0428-type-inheritance-and-forking.md

# Reference-level explanation

[reference-level-explanation]: #reference-level-explanation

## Defining extended entity types

In the Block Protocol, we will allow type extension through the `allOf` JSON Schema keyword which specifies an array of schemas that will have to validate together.

We'll add the following fields to the existing Entity Type meta schema definition:

```json
{
  "$id": "https://blockprotocol.org/type-system/0.3/schema/meta/entity-type",
  "type": "object",
  ...,
  "properties": {
    ...,
    "allOf": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "$ref": {
            "$comment": "Valid reference to an existing Entity Type version",

>             "$ref": "$ref": "https://blockprotocol.org/types/modules/graph/0.3/schema/versioned-url"

Reviewer: ```suggestion
            "$ref": "https://blockprotocol.org/types/modules/graph/0.3/schema/versioned-url"
```

Author: Thanks! My vim bindings likely screwed that one up..

Reviewer: No worries 😄

- Block Protocol Discussion: [blockprotocol/blockprotocol#439](https://github.com/blockprotocol/blockprotocol/discussions/439)

# Summary

[summary]: #summary

As a follow-up to the [Graph type system RFC](./0352-graph-type-system.md), this RFC will describe the behavior of which types in the type system can be extended and duplicated to enhance reusability while giving more control to users (both block authors and users of embedding applications). Extending types can be considered the same as "type inheritance", but there are some important nuances that make these concepts different.

Type duplication allows for modifying a type completely without restrictions on how, at the expense of the guarantees that can be made for type extension. Duplication can be considered the same as "forking".

# Motivation

[motivation]: #motivation

Reusing public types in the type system comes with the potential disadvantage of not fully conforming to a user's intention. If a user is interested in a public type but needs certain additional properties to make the type usable for their use case, they would have to recreate the type themself in the current system.

> The types in the type systems all specify JSON Schema's `{ "additionalProperties": false }` which has implications for how the schemas can be composed together, which will be addressed in this proposal.

Author: Reworded slightly here 04746a571a16c228326bf212efe2af440dd37068

For example, an `Employee` instance looks as follows (simplified):

```json
{
  "name": "Charles",
  "age": 35,
  "occupation": "Merchant"
}
```

And the properties would have the following relation to `Employee`:

```txt
      name◄┐ (supertype)    (subtype)
           ├───Person───────Employee
       age◄┘                │
                            │
occupation◄─────────────────┘
```

We can visually see how selecting `Person` in the type hierarchy would provide `name` and `age` properties but exclude the `occupation` property.
Assuming that we can project/select the properties of a type that are defined through the supertype, coercive subtyping is attainable for any subtype. This is a somewhat strong assumption to make, but it unlocks expressing how to extend types. The reason the selection/projection of properties is important for coercive subtyping will be explained in more detail in the coming sections, specifically in the [`additionalProperties` problem explanation section](#the-additionalproperties-problem).

> While it was mentioned that we don't allow overriding properties, we do allow it to a certain extent. When properties are still _compatible_ (using the definition of compatible given in the [versioning RFC](./0408-versioning-types.md#determining-type-compatibility)), properties can be overridden. For example, if `Person` had an optional property `name`, `Employee` could override it to be a required property. This is because the property is still compatible, but it is now more specific. This is also the reason why we can't allow overriding properties to be arrays or objects, as that would make the property incompatible.

Reviewer: Maybe we should change the wording slightly:
```suggestion
While it was mentioned that we don't allow overriding properties, we do allow modifications to a certain extent. When properties are still _compatible_ (using the definition of compatible given in the [versioning RFC](./0408-versioning-types.md#determining-type-compatibility)), properties can be overridden. For example, if `Person` had an optional property `name`, `Employee` could override it to be a required property. This is because the property is still compatible, but it is now more specific. This is also the reason why we can't allow overriding properties to be arrays or objects, as that would make the property incompatible.
```

The assumption that we can select/project parts of a subtype that make up a supertype is essential for keeping strictness in JSON Schemas.

We propose slight modifications to how `{ "additionalProperties": true }` and `{ "unevaluatedProperties": false }` behave and may be used within our type extension system to let supertypes keep strictness while allowing composition. This modification is a semantic change and does not change any syntax.

Concrete examples of how JSON Schema breaks with these validation constraints are shown in the [Reference-level explanation](#problems-with-unevaluatedproperties).

## Defining extended entity types

Extended types will be defined with conventional JSON Schema syntax: the `allOf` keyword. An entity type can extend another entity type by adding a versioned URI reference to the root-level `allOf` array.

> A [versioned URI](https://github.com/blockprotocol/blockprotocol/blob/main/rfcs/text/0408-versioning-types.md#type-uris) is used so that subtypes aren't automatically updated (and potentially invalidated) when the supertype is updated.

Reviewer: ```suggestion
A [versioned URL](https://github.com/blockprotocol/blockprotocol/blob/main/rfcs/text/0408-versioning-types.md#type-uris) is used so that subtypes aren't automatically updated (and potentially invalidated) when the supertype is updated.
```

Author: Swept through URI in 3615bb1848ac4250f017197cfc1aa2536f424df0 :)

Reviewer: Thanks!

> As extended types can extend other extended types we must also make sure that there are no cycles within the type hierarchy, as it makes types difficult to resolve/reason about and could lead to unpredictable behavior.

Reviewer: Clarification question: For cycles being possible, this would require inserting (or update) multiple types at the same time, right? Wouldn't this imply that types would be incompatible?

Author: I think that would be a situation where cycles could lead to incompatibility, yes. I guess that would fall into the category of unpredictable behavior. Since the type system doesn't really allow "transactions" or predicting the version URLs, I think we're covered for now. If we introduce anything like that we have to be careful with inheritance though

Reviewer: Thank you 🙂

```txt
name◄─────Being───┐
                  │
          age◄────Person─────Employee
                             │
   occupation◄───────────────┘
```

where `Employee` is a subtype of `Person` which in turn is a subtype of `Being`. The three properties `name`, `age` and `occupation` would be present on `Employee` and considered all expanded properties for `Employee`.
The user wants to change the `age` property to a `tenure` property through duplication. They would have to create a new subtype of `Being` with the desired changes:

```txt
name◄─────Being───┐
                  │
       tenure◄────MyEmployee
                  │
   occupation◄────┘
```

From the perspective of the user duplicating a type, the `Person` and `Employee` entity types were squashed together or "expanded" for duplication to have the desired effect. The `Being` supertype relation is still kept when changing a property on the `MyEmployee` duplicated type. But in this process, the `MyEmployee` entity type loses the ability to trivially be treated as a `Person`, as there is no subtyping relation with `Person` on the new `MyEmployee` duplicated type.

> The duplicatio nhappened on the expanded type resulting in combining/expanding `Person` and `Employee`, but it would also have been possible expand the entire type, getting rid of `Being`. This would have resulted in the following type hierarchy:

Reviewer: ```suggestion
The duplication happened on the expanded type resulting in combining/expanding `Person` and `Employee`, but it would also have been possible to expand the entire type, getting rid of `Being`. This would have resulted in the following type hierarchy:
```

```txt
      name◄────┐
               │
    tenure◄────MyEmployee
               │
occupation◄────┘
```

> And like the previous example, the `MyEmployee` type would not be a subtype of `Person` anymore. It also wouldn't be a subtype of `Being` either - reusability of the types is lost in this case. Ideally type duplcation would be able to preserve as many supertype relations as possible.

Reviewer: ```suggestion
And like the previous example, the `MyEmployee` type would not be a subtype of `Person` anymore. It also wouldn't be a subtype of `Being` either - the reusability of the types is lost in this case. Ideally, type duplication would be able to preserve as many supertype relations as possible.
```

# Reference-level explanation

[reference-level-explanation]: #reference-level-explanation

## Defining extended entity types

In the Block Protocol, we will allow type extension through the `allOf` JSON Schema keyword which specifies an array of schemas that will have to validate together.

We'll add the following fields to the existing Entity Type meta schema definition:

```json
{
  "$id": "https://blockprotocol.org/type-system/0.3/schema/meta/entity-type",
  "type": "object",
  ...,
  "properties": {
    ...,
    "allOf": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "$ref": {
            "$comment": "Valid reference to an existing Entity Type version",

>             "type": "string",
>             "format": "uri"

Reviewer: ```suggestion
            "$ref": "https://blockprotocol.org/types/modules/graph/0.3/schema/versioned-url"
```

          },
          "additionalProperties": false
        }
      }
    }
  }
}
```

### Concrete examples of extended entity types

**An example of _disjoint_ properties**:

Given a _supertype_ `Person`:

```json

> {

Reviewer: ```suggestion
{
  "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/entity-type",
```

Reviewer: Could you please search for other occurrences and add the "$schema" there as well?

Author: Yes of course! Should be all here b6f59c160a55d58ddd446c476d4d7b9310ec8a42

        "$ref": "https://blockprotocol.org/@alice/property-type/phone-number/v/1"
      },
      // The update is inlined in the existing schema
      "https://blockprotocol.org/@alice/property-type/birth-date": {
        "$ref": "https://blockprotocol.org/@alice/property-type/birth-date/v/1"
      }
    }
  }
}
```

given that the original schema was created as follows

```json
{
  "$id": "https://blockprotocol.org/@alice/entity-type/person/v/1",
  "type": "object",
  "kind": "entityType",
  "title": "Person",
  "properties": {
    "https://blockprotocol.org/@alice/property-type/name": {
      "$ref": "https://blockprotocol.org/@alice/property-type/name/v/1"
    },
    "https://blockprotocol.org/@alice/property-type/email": {
      "$ref": "https://blockprotocol.org/@alice/property-type/email/v/1"
    },
    "https://blockprotocol.org/@alice/property-type/phone-number": {
      "$ref": "https://blockprotocol.org/@alice/property-type/phone-number/v/1"
    }
  }
}
```

The above change applies to these Block Protocol operations (`updateEntityType` having already been implemented, and the others proposed in accepted RFCs):

- `updateEntityType`
- `updatePropertyType`

> - `updateLinkType`

Reviewer: ```suggestion
```

which must all make use of complete schemas in place of partial ones. These examples originate from the [Type System RFC](./0352-graph-type-system.md#interfacing-with-types-1).

> The `aggregateEntities` operation may need to change its behavior slightly. When calling `aggregateEntities` for a specific entity type, entities that are returned might be of that entity type or subtypes of it when allowing extended types. This is because the entities of the subtype can be coerced to the requested entity type.

Reviewer: ```suggestion
The `queryEntities` operation may need to change its behavior slightly. When calling `queryEntities` for a specific entity type, entities that are returned might be of that entity type or subtypes of it when allowing extended types. This is because the entities of the subtype can be coerced to the requested entity type.
```
